apps/consulting/posts/review-functorch-contribution.md

```diff
@@ -0,0 +1,327 @@
+their ability to compose with one another. Thanks to this, the process of
+calculating per-sample gradients becomes the straightforward application of
+`vmap(grad(model))`.'
+---
+
+<base target="_blank" />
+
+`torch.func` (previously known as `functorch`) is a PyTorch module designed to
+offer JAX-like transforms. Within this module, various higher-order
+functions, such as `grad`, `vmap`, and `vjp` are made accessible. These
+transforms help users to easily compute gradients for the parameters of their model
+or write batch-size agnostic code. The beauty of these transformations lies in
+their ability to compose with one another. Thanks to this, the process of
+calculating per-sample gradients becomes the straightforward application of
+`vmap(grad(model))`.
+
+Here are a few of the tasks we had the opportunity to tackle:
+
+## Adding Batching Rule for `vmap`
+`vmap` is a transformation that accepts a function that operates on non-batched
+data and returns a new function that operates on batched input.
+When processing a batched input, an additional
+dimension, denoted by `in_dims`, is introduced to indicate which dimension to
+apply the function over. Conceptually, it emulates a `for` loop that iterates
+through all the data points and stacks the results. Importantly, it performs this
+operation efficiently by pushing the `for` loop into the PyTorch operations,
+effectively vectorizing the computation.
+
+Consider the following example:
+```python
+import torch
+
+# Written to handle only single sample.
+def my_simple_model(feature_vec, weight):
+    return torch.dot(feature_vec, weight).relu()
+
+batch_size = 4
+batched_inputs = torch.randn(batch_size, 3)
+weight = torch.randn(3)
+
+# For Loop version
+expected = []
+for input in batched_inputs:
+    expected.append(my_simple_model(input, weight))
+expected = torch.stack(expected)
+
+# Vmap
+# `in_dims` specifies the dimension that should be mapped over.
+# In this case, we map only over 0-dim of `batched_inputs`.
+actual = torch.vmap(my_simple_model, in_dims=(0, None))(batched_inputs, weight)
+
+# Verify that the results match.
+torch.testing.assert_close(expected, actual)
+```
+
+To support `vmap` for PyTorch operators, we need to specify the batching rule
+i.e. how to map the given function over a batched input.
+A batching rule is essentially a function which takes one or multiple batched
+inputs and computes the batched operation. In the above example to support
+`vmap` for `my_simple_model`, we need to know the batching rule for `torch.dot`
+and `torch.relu` to be able to vectorize our model.
+PyTorch has a lot of operators and we need to have coverage for all of them
+to support `vmap`. That being said, there is a `for`-loop
+fallback in case an operator is not supported so as not to crash the code.
+
+PyTorch operators can be roughly categorized as primitive (internally
+`CompositeExplicitAutograd`) vs composite (internally `CompositeImplicitAutograd`).
```

Comment:
This distinction is confusing since both are composite. It's also inaccurate since most core operators are probably `CPU, CUDA` dispatch.

Reply:
Hmm, agreed. I think we should avoid using these internal terms then. How does the following sound?

> From the point of view of adding batch rule, PyTorch operators can be roughly categorised as primitive and composite.

Reply:
NIT: "primitive or composite" othewise LGTM